pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -103,206 +141,309 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
-		close(e.stopChan)
-		e.stopChan = nil
+	// Spawn workers to process work queues
+	for i := 0; i < workers; i++ {
+		go wait.Until(e.syncServiceAccount, 0, stopCh)
+		go wait.Until(e.syncSecret, 0, stopCh)
 	}
+
+	// Block until stop channel is closed
+	<-stopCh
+
+	// Shut down queues
+	e.syncServiceAccountQueue.ShutDown()
+	e.syncSecretQueue.ShutDown()
 }
 
-// serviceAccountAdded reacts to a ServiceAccount creation by creating a corresponding ServiceAccountToken Secret
-func (e *TokensController) serviceAccountAdded(obj interface{}) {
-	serviceAccount := obj.(*api.ServiceAccount)
-	err := e.createSecretIfNeeded(serviceAccount)
-	if err != nil {
-		glog.Error(err)
+func (e *TokensController) queueServiceAccountSync(obj interface{}) {
+	if serviceAccount, ok := obj.(*api.ServiceAccount); ok {
+		e.syncServiceAccountQueue.Add(makeServiceAccountKey(serviceAccount))
 	}
 }
 
-// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
-func (e *TokensController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
-	newServiceAccount := newObj.(*api.ServiceAccount)
-	err := e.createSecretIfNeeded(newServiceAccount)
-	if err != nil {
-		glog.Error(err)
+func (e *TokensController) queueServiceAccountUpdateSync(oldObj interface{}, newObj interface{}) {
+	if serviceAccount, ok := newObj.(*api.ServiceAccount); ok {
+		e.syncServiceAccountQueue.Add(makeServiceAccountKey(serviceAccount))
 	}
 }
 
-// serviceAccountDeleted reacts to a ServiceAccount deletion by deleting all corresponding ServiceAccountToken Secrets
-func (e *TokensController) serviceAccountDeleted(obj interface{}) {
-	serviceAccount, ok := obj.(*api.ServiceAccount)
-	if !ok {
-		// Unknown type. If we missed a ServiceAccount deletion, the
-		// corresponding secrets will be cleaned up during the Secret re-list
+func (e *TokensController) retry(queue *workqueue.Type, key interface{}) {
+	// get retry count for this key
+	retryCount := 0
+	if c, ok := e.retryCounts.Get(key); ok {
+		retryCount = c.(int)
+	}
+
+	// stop retrying after N attempts
+	if retryCount >= e.maxRetries {
+		glog.V(4).Infof("retried %d times: %#v", retryCount, key)
 		return
 	}
-	secrets, err := e.listTokenSecrets(serviceAccount)
-	if err != nil {
-		glog.Error(err)
+
+	// increment our retry count
+	retryCount++
+	e.retryCounts.Add(key, retryCount)
+
+	// See how long we should wait before retrying
+	delay := e.retryRateLimiter.Take(1)
+
+	// Retry immediately
+	if delay == 0 {
+		queue.Add(key)
 		return
 	}
-	for _, secret := range secrets {
-		glog.V(4).Infof("Deleting secret %s/%s because service account %s was deleted", secret.Namespace, secret.Name, serviceAccount.Name)
-		if err := e.deleteSecret(secret); err != nil {
-			glog.Errorf("Error deleting secret %s/%s: %v", secret.Namespace, secret.Name, err)
-		}
+
+	// Or delay
+	glog.V(4).Infof("rate-limiter delay of %s retrying %#v", key, delay)
+	go func() {
```

Comment:
I think the fanout here is excessive, especially in cases of environmental failures happening on initial activation.  I'd still say that https://github.com/kubernetes/kubernetes/pull/23444 is a more appropriate place to perform the requeue action and probably a better place to do the overall rate limiting too.